### PR TITLE
[DeviceIndex][7/N] Use DeviceIndex in XPU

### PR DESCRIPTION
**Pull request by cyyever**

### CODE REVIEW

File: c10/xpu/XPUFunctions.cpp [hunk 1]
```diff
@@ -146,8 +144,9 @@ int get_device_idx_from_pointer(void* ptr) {
       gDevicePool.devices.begin(), gDevicePool.devices.end(), match_device);
   TORCH_CHECK(
       it != gDevicePool.devices.end(),
-      "Cant't find the pointer from XPU devices.");
-  return static_cast<int>(std::distance(gDevicePool.devices.begin(), it));
+      "Can't find the pointer from XPU devices.");
+  auto device_idx = std::distance(devices.begin(), it);
+  return static_cast<DeviceIndex>(std::distance(devices.begin(), it));
```
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:149]:
```suggestion
  return static_cast<DeviceIndex>(std::distance(gDevicePool.devices.begin(), it));
```

File: aten/src/ATen/DLConvertor.cpp [hunk 2]
```diff
@@ -88,9 +88,9 @@ DLDataType getDLDataType(const Tensor& t) {
   return dtype;
 }
 
-static DLDevice getDLDevice(const Tensor& tensor, const int64_t& device_id) {
+static DLDevice getDLDevice(const Tensor& tensor, c10::DeviceIndex device_id) {
   DLDevice ctx;
-  ctx.device_id = device_id;
+  ctx.device_id = static_cast<int>(device_id);
```
Comment (guangyey) [inline aten/src/ATen/DLConvertor.cpp:93]: Explicitly  `int32_t` looks better?
`device_id` in DLDevice is a `int32_t`, see https://github.com/pytorch/pytorch/blob/f4cf25bb24be735b2502ae13f290017992c2fac8/aten/src/ATen/dlpack.h#L97-L105

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 3]
```diff
@@ -23,7 +21,7 @@ std::deque<c10::once_flag> device_prop_flags;
 std::vector<DeviceProp> device_properties;
 
 std::deque<c10::once_flag> device_global_idx_flags;
-std::vector<int32_t> device_global_idxs;
+std::vector<c10::DeviceIndex> device_global_idxs;
```
Comment (guangyey) [inline aten/src/ATen/xpu/XPUContext.cpp:24]:
```suggestion
std::vector<int32_t> device_global_idxs;
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 4]
```diff
@@ -46,9 +44,9 @@ void initDeviceGlobalIdx(DeviceIndex device) {
   };
   auto it = std::find_if(devices.begin(), devices.end(), match_device);
   TORCH_CHECK(
-      it != devices.end(), "Cant't find the global index of XPU device.");
+      it != devices.end(), "Can't find the global index of XPU device.");
   device_global_idxs[device] =
-      static_cast<int32_t>(std::distance(devices.begin(), it));
+      static_cast<c10::DeviceIndex>(std::distance(devices.begin(), it));
```
Comment (guangyey) [inline aten/src/ATen/xpu/XPUContext.cpp:49]:
```suggestion
      static_cast<int32_t>(std::distance(devices.begin(), it));
```

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 5]
```diff
@@ -79,7 +77,7 @@ DeviceProp* getDeviceProperties(DeviceIndex device) {
 
 // Return the global index enumerated by sycl::device::get_devices based on the
 // index of a XPU device in the framework.
-int32_t getGlobalIdxFromDevice(DeviceIndex device) {
+c10::DeviceIndex getGlobalIdxFromDevice(DeviceIndex device) {
```
Comment (guangyey) [inline aten/src/ATen/xpu/XPUContext.cpp:80]: This API is used in dlpack and here returns a global index enumerated by `sycl::device::get_devices` which contains XPU devices and other devices. And the return value may be larger than the number of XPU devices. So it is possible to result in an overflow if this API is defined in `DeviceIndex`. We should align the return type with dlpack using `int32_t` the same as `device_id` defined in `DLDevice`.

Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp:80]:
```suggestion
int32_t getGlobalIdxFromDevice(DeviceIndex device) {
```

File: aten/src/ATen/xpu/XPUContext.h [hunk 6]
```diff
@@ -15,6 +15,6 @@ TORCH_XPU_API DeviceProp* getCurrentDeviceProperties();
 
 TORCH_XPU_API DeviceProp* getDeviceProperties(DeviceIndex device);
 
-TORCH_XPU_API int32_t getGlobalIdxFromDevice(DeviceIndex device);
+TORCH_XPU_API c10::DeviceIndex getGlobalIdxFromDevice(DeviceIndex device);
```
Comment (guangyey) [inline aten/src/ATen/xpu/XPUContext.h:18]:
```suggestion
TORCH_XPU_API int32_t  getGlobalIdxFromDevice(DeviceIndex device);
```

File: aten/src/ATen/xpu/detail/XPUHooks.cpp [hunk 7]
```diff
@@ -21,7 +21,8 @@ std::string XPUHooks::showConfig() const {
   return "XPU backend";
 }
 
-int XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
+c10::DeviceIndex XPUHooks::getGlobalIdxFromDevice(
```
Comment (guangyey) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:24]:
```suggestion
int32_t  XPUHooks::getGlobalIdxFromDevice(
```

File: aten/src/ATen/xpu/detail/XPUHooks.h [hunk 8]
```diff
@@ -13,9 +10,10 @@ struct XPUHooks : public at::XPUHooksInterface {
   void initXPU() const override;
   bool hasXPU() const override;
   std::string showConfig() const override;
-  int getGlobalIdxFromDevice(const at::Device& device) const override;
+  c10::DeviceIndex getGlobalIdxFromDevice(
```
Comment (guangyey) [inline aten/src/ATen/xpu/detail/XPUHooks.h:13]:
```suggestion
  int32_t getGlobalIdxFromDevice(const at::Device& device) const override;
```

File: aten/src/ATen/xpu/detail/XPUHooks.h [hunk 9]
```diff
@@ -13,9 +10,10 @@ struct XPUHooks : public at::XPUHooksInterface {
...
+      const at::Device& device) const override;
```
Comment (guangyey) [inline aten/src/ATen/xpu/detail/XPUHooks.h:14]:
```suggestion
  int32_t getGlobalIdxFromDevice(const at::Device& device) const override;
```

File: aten/src/ATen/detail/XPUHooksInterface.h [hunk 10]
```diff
@@ -39,7 +39,7 @@ struct TORCH_API XPUHooksInterface {
         XPU_HELP);
   }
 
-  virtual int getGlobalIdxFromDevice(const Device& device) const {
+  virtual DeviceIndex getGlobalIdxFromDevice(const Device& device) const {
```
Comment (guangyey) [inline aten/src/ATen/detail/XPUHooksInterface.h:42]:
```suggestion
  virtual int32_t getGlobalIdxFromDevice(const Device& device) const {
```

Comment (guangyey) [reply aten/src/ATen/detail/XPUHooksInterface.h:42]: This API in hooks is used in dlpack.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp:24]: for dlpack

File: aten/src/ATen/xpu/detail/XPUHooks.cpp (hunk 7 above)
Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:24]: for dlpack

**Comment by guangyey** [comment]
LGTM, thank you for your help.

See a minor comments: 
`getGlobalIdxFromDevice` is used in dlpack. Its data type of return value needs to align with `device_id` defined in dlpack.h. So it is better to use `int32_t`.

**Review by guangyey — APPROVED** [review]

File: aten/src/ATen/detail/XPUHooksInterface.h (hunk 10 above)
Comment (cyyever) [reply aten/src/ATen/detail/XPUHooksInterface.h:42]: dlpack was changed to this new API in the PR

**Comment by cyyever** [comment]
@guangyey Thank you for the informative review. When changed the type of global device index, I noted that the pool is initialised by a size of DeviceIndex, so there is no danger of integer overflow.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 5 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp:80]: `device_global_idxs`'s size is a `DeviceIndex` type. But the value stored in `device_global_idxs` maybe larger than `DeviceIndex`, which results in an overflow, like below:
```cpp
device_global_idxs.size() = 5; // a `DeviceIndex type, 5 xpu devices in this machine.
device_global_idxs[0] = 100000; // the 0th device's global id is 100000, resulting in an overflow.
```

**Review by guangyey — APPROVED** [review]
Thanks for your update!

**Review by Skylion007 — APPROVED** [review]

File: torch/xpu/__init__.py [hunk 11]
```diff
@@ -382,7 +382,7 @@ def synchronize(device: _device_t = None) -> None:
             if :attr:`device` is ``None`` (default).
     """
     _lazy_init()
-    device = _get_device_index(device)
+    device = _get_device_index(device, optional=True)
```
Comment (guangyey) [inline torch/xpu/__init__.py:385]: Thanks @cyyever

**Pull request closed without merging**